docs/record_reader.rst

```diff
@@ -0,0 +1,133 @@
+.. "License"); you may not use this file except in compliance
+.. with the License.  You may obtain a copy of the License at
+..
+..   http://www.apache.org/licenses/LICENSE-2.0
+..
+.. Unless required by applicable law or agreed to in writing,
+.. software distributed under the License is distributed on an
+.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+.. KIND, either express or implied.  See the License for the
+.. specific language governing permissions and limitations
+.. under the License.
+..
+
+Record Reader
+=============
+
+Pinot supports indexing data from various file formats. To support reading from a file format, a record reader need to
+be provided to read the file and convert records into the general format which the indexing engine can understand. The
+record reader serves as the connector from each individual file format to Pinot record format.
+
+Pinot package provides the following record readers out of the box:
+
+- Avro record reader: record reader for Avro format files
+- CSV record reader: record reader for CSV format files
+- JSON record reader: record reader for JSON format files
+- Thrift record reader: record reader for Thrift format files
+- Pinot segment record reader: record reader for Pinot segment
+
+For other file formats, we provide a general interface for record reader - `RecordReader`. To index the file into Pinot
+segment, implement the interface and plug it into the index engine - `SegmentCreationDriverImpl`.
+
+Initialize Record Reader
+------------------------
+
+To initialize a record reader, the data file and table schema should be provided (for Pinot segment record reader, only
+need to provide the index directory because schema can be derived from the segment). The output record will follow the
+table schema provided.
+
+For Avro/JSON/Pinot segment record reader, no extra configuration is required as column names and multi-values are
+embedded in the data file.
+
+For CSV/Thrift record reader, extra configuration might be provided to determine the column names and multi-values for
+the data.
+
+CSV Record Reader Config
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+The CSV record reader config contains the following settings:
+
+- Header: the header for the CSV file (column names)
+- Column delimiter: delimiter for each column
+- Multi-value delimiter: delimiter for each value for a multi-valued column
+
+If no config provided, use the default setting:
+
+- Use the first row in the data file as the header
+- Use ',' as the column delimiter
+- Use ';' as the multi-value delimiter
+
+Thrift Record Reader Config
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The Thrift record reader config is mandatory. It contains the Thrift class name for the record reader to de-serialize
+the Thrift objects.
+
+Implement Your Own Record Reader
+--------------------------------
+
+You can implement your own record reader for file formats that is not supported natively. The following methods need to
```

Comment:
I would prefer if we just point to the record reader interface in the code; else we'll have to update this doc everytime there is *any* change to the interface.

- I would put explanations for GenericRow etc before pointing to the interface.
- It will also be good to high light that we use a 2 pass algorithm, hence "rewind" is necessary. This will inform the readers about additional restrictions on the interface.

Reply:
Done